File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/templatemethod/transport/Transport.java
// Transport.java

package com.bigboxcode.designpattern.templatemethod.transport;

public abstract class Transport {
    abstract void createBody();

    abstract void addEngine();

    abstract void addWheel();

    // Required only for Plane
    abstract void addWing();

    private void addSeat() {
        // Add seats to the vehicle

        // Adding seats are same for all transports so same functions for all

        System.out.println("Adding seats");
    }

    private void paint() {
        System.out.println("Painting");
    }

    public void build() {
        createBody();

        addEngine();

        addWheel();

        addWing();

        addSeat();

        paint();
    }
}

File: design-patterns/ts/template-method/transport/transport.ts
// transport.ts

abstract class Transport {
    public abstract createBody(): void;
    public abstract addEngine(): void;
    public abstract addWheel(): void;
    public abstract addWing(): void;

    private addSeat(): void {
        console.log("Adding seats");
    }

    private paint(): void {
        console.log("Painting");
    }

    public build(): void {
        this.createBody();
        this.addEngine();
        this.addWheel();
        this.addWing();
        this.addSeat();
        this.paint();
    }
}

export default Transport;